{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from itertools import chain\n",
    "import xlwt\n",
    "from Bio import SeqIO\n",
    "from snapgene_reader import snapgene_file_to_dict, snapgene_file_to_seqrecord\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "class plasmidIDtool(): \n",
    "    \"\"\"\n",
    "    Tool to search sequence of DNA and find regions of interest\n",
    "    Support only .seq files\n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        \n",
    "        self.wd = ''\n",
    "        self.lis =[]\n",
    "        self.region = ''\n",
    "        self.searchRange = 0\n",
    "        self.searchRegionLen = 0\n",
    "        self.name =''\n",
    "        self.complete = False \n",
    "        self.filename = ''\n",
    "             \n",
    "            \n",
    "    def fetch_wd(self,wd): \n",
    "        \"\"\"\n",
    "        verify working dir exsists\n",
    "        \"\"\"\n",
    "        try:\n",
    "            os.chdir(wd)\n",
    "        except OSError:\n",
    "            print \"\"\"Error: Sorry can't find this folder, recheck path\"\"\"\n",
    "            return False\n",
    "            \n",
    "        else:\n",
    "            print \"Written content in the file successfully\"\n",
    "            self.wd = wd \n",
    "            return True\n",
    "        \n",
    "        \n",
    "    def check_files(self):\n",
    "        \n",
    "        \"\"\"\n",
    "        generate list of .seq files \n",
    "        \"\"\"\n",
    "        self.lis = []\n",
    "        direc = os.listdir(self.wd)\n",
    "        \n",
    "        for file_ in direc: \n",
    "            if file_.endswith(('.gb','.fa','.dna','.seq')): self.lis.append(file_)\n",
    "                \n",
    "        return self.lis \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    def ProduceCleanSeq(self,filename): \n",
    "        #MAKE SURE USELESS IS A CONSTANT \n",
    "        \"\"\"\n",
    "        Format file, removing unessential characters. Useless constant is the character length of \n",
    "        the file for .seq,.fa,.gb,.dna files.\n",
    "    \n",
    "        \"\"\"\n",
    "       \n",
    "        if filename.endswith('.fa'):\n",
    "            for seq_record in SeqIO.parse(filename, \"fasta\"): Data_formated = str(seq_record.seq).upper()\n",
    "        elif filename.endswith('.gb'):\n",
    "            for seq_record in SeqIO.parse(filename, \"genbank\"): Data_formated = str(seq_record.seq).upper()    \n",
    "        elif filename.endswith('.dna'): \n",
    "            Data_formated = str(snapgene_file_to_dict(filename)['seq']).upper()\n",
    "        elif filename.endswith('.seq'):\n",
    "                useless = 12 \n",
    "                data = open(filename).read().split()\n",
    "                Data_unformatted = list(chain.from_iterable(data[i] for i in range(len(data))))\n",
    "                Data_formated = Data_unformatted[useless:]\n",
    "                \n",
    "        Data_str = ''.join(Data_formated)\n",
    "        Clean_Seq = ''.join([i for i in Data_str if not i.isdigit()])\n",
    "        \n",
    "        return Clean_Seq\n",
    "\n",
    "\n",
    "    def find_region(self,Region,clean_seq):\n",
    "        \"\"\"\n",
    "        return bool if region is in sequence\n",
    "        \"\"\"\n",
    "       \n",
    "        return  Region in clean_seq\n",
    "    \n",
    "    \n",
    "    \n",
    "    def find_region_range(self,Region,clean_seq,searchRange,searchRegionLen): \n",
    "        #DEAL WITH THE WRAPAROUND ISSUE \n",
    "        \n",
    "        \"\"\"\n",
    "        Look for given region in the cleaned sequence\n",
    "        Foreward is 3' to 5'\n",
    "        \"\"\"\n",
    "        \n",
    "        \n",
    "        if self.find_region(Region,clean_seq) == False: \n",
    "            print \"Region does not exsist in given sequence \"\n",
    "            print \"  \"\n",
    "            new_regF, new_regB = \"NaN\",\"NaN\"\n",
    "        \n",
    "        else:\n",
    "            print \"Region Found\"\n",
    "            print \"  \"\n",
    "            regionlen = len(Region)\n",
    "            region_start = clean_seq.find(Region)\n",
    "            region_end = region_start + regionlen\n",
    "            new_pos_F = region_end + searchRange\n",
    "            new_pos_B = region_start - searchRange\n",
    "            \n",
    "            #deal with wrap around issue. \n",
    "            if new_pos_B < 0: \n",
    "                #THINK OF BETTER WAY TO DO THIS\n",
    "                new_pos_B = len(clean_seq) * 2\n",
    "            \n",
    "            new_regF = clean_seq[new_pos_F : new_pos_F + searchRegionLen]\n",
    "            new_regB = clean_seq[new_pos_B - searchRegionLen: new_pos_B]\n",
    "\n",
    "        return new_regF, new_regB\n",
    "\n",
    "   \n",
    "\n",
    "    \n",
    "    def rev_comp(self,seq): \n",
    "        \"\"\"\n",
    "        Produce the reverse complement sequence of given sequence\n",
    "        \"\"\"\n",
    "        if seq == \"NaN\": rev_comp = \"NaN\"\n",
    "        else:     \n",
    "            complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "            rev_comp = ''.join([complement[base] for base in seq[::-1]])\n",
    "        \n",
    "        return rev_comp\n",
    "    \n",
    "    def DNA_to_protein(self,seq): \n",
    "        \"\"\"\n",
    "        Convert DNA to Amino acid\n",
    "        \"\"\"\n",
    "        if seq == \"NaN\": protein = \"NaN\" \n",
    "        elif \"N\" in seq: protein = \"Invalid sequence\"\n",
    "       \n",
    "        else:\n",
    "            table = {\n",
    "                'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "                'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "                'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "                'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',                 \n",
    "                'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "                'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "                'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "                'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "                'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "                'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "                'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "                'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "                'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "                'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "                'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "                'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "            }\n",
    "\n",
    "            protein =\"\"\n",
    "            if len(seq)%3 == 0:\n",
    "                for i in range(0, len(seq), 3):\n",
    "                    codon = seq[i:i + 3]\n",
    "                    protein+= table[codon]\n",
    "        return protein \n",
    "\n",
    "    \n",
    "    def start(self): \n",
    "        \n",
    "        \"\"\"\n",
    "        Runs tool UI\n",
    "        \"\"\"\n",
    "        while self.complete == False:\n",
    "            \n",
    "            #Enter the path name by user \n",
    "            wd = str(raw_input('enter the folder path you would like to analyze: '))\n",
    "            \n",
    "            if self.fetch_wd(wd) == True: \n",
    "                print 'Checking files'\n",
    "                self.check_files()\n",
    "                print \"You have\" + \" \" + str(len(self.lis)) + \" \" + \"file(s)\"\n",
    "                print self.lis\n",
    "            else: \n",
    "                print \"  \"\n",
    "                print \"Cant Find you File, lets try again\"\n",
    "            \n",
    "            if len(self.lis) == 0: print \"nothing in here, lets try again\"    \n",
    "            else:\n",
    "                #Enter specifics for user\n",
    "                #self.region = str(raw_input('enter the refernce region you would like to search: '))\n",
    "                self.region = (str(raw_input('enter the refernce region(s) you would like to search, seperate regions with space: '))).split()\n",
    "                self.searchRange = int(raw_input('how many base pairs would you like to search : '))\n",
    "                self.searchRegionLen = int(raw_input('how many base pairs is region you like to find : '))\n",
    "                self.name = raw_input('what you like to call your excel file : ')\n",
    "                \n",
    "                #Generate Excel files \n",
    "                 #DONT NEED BACKWARD SEARCHER \n",
    "                row = 0\n",
    "                book = xlwt.Workbook(encoding=\"utf-8\")\n",
    "                sheet1 = book.add_sheet(\"Data\")\n",
    "                sheet1.write(0, 0, \"File Name\")\n",
    "                sheet1.write(0, 1, \"Reference Region\")\n",
    "                sheet1.write(0, 2, \"Search Distance\")\n",
    "                sheet1.write(0, 3, \"New Region Foreward\")   \n",
    "                sheet1.write(0, 4, \"New Region Backward\")\n",
    "                sheet1.write(0, 5, \"Protein Foreward\")\n",
    "                sheet1.write(0, 6, \"Protein Backward\")\n",
    "                sheet1.write(0, 7, \"New Region Foreward(Rev Comp)\")\n",
    "                sheet1.write(0, 8, \"New Region Backward(Rev Comp)\")\n",
    "                sheet1.write(0, 9, \"Protein Foreward(Rev Comp)\")\n",
    "                sheet1.write(0, 10, \"Protein Backward(Rev Comp)\")\n",
    "                \n",
    "                for reg in self.region:\n",
    "                    for idx in range(len(self.lis)):\n",
    "                        print self.lis[idx]\n",
    "                        #Correct for row in excel\n",
    "                        row += 1 \n",
    "                       \n",
    "                        #Clean up sequence \n",
    "                        clean_seq = self.ProduceCleanSeq(self.lis[idx])\n",
    "\n",
    "                        #Find regions and convert to proteins \n",
    "                        new_regF, new_regB = self.find_region_range(reg,clean_seq,self.searchRange,self.searchRegionLen)\n",
    "                        proteinF = self.DNA_to_protein(new_regF)\n",
    "                        proteinB = self.DNA_to_protein(new_regB)\n",
    "                        new_regF_rc = self.rev_comp(new_regF)\n",
    "                        new_regB_rc = self.rev_comp(new_regB)\n",
    "                        proteinF_rc = self.DNA_to_protein(new_regF_rc)\n",
    "                        proteinB_rc = self.DNA_to_protein(new_regB_rc)\n",
    "\n",
    "                        #enter in the file \n",
    "                        sheet1.write(row, 0, self.lis[idx])\n",
    "                        sheet1.write(row, 1, reg)\n",
    "                        sheet1.write(row, 2, self.searchRange)\n",
    "                        sheet1.write(row, 3, new_regF)\n",
    "                        sheet1.write(row, 4, new_regB)\n",
    "                        sheet1.write(row, 5, proteinF)\n",
    "                        sheet1.write(row, 6, proteinB)\n",
    "                        sheet1.write(row, 7, new_regF_rc)\n",
    "                        sheet1.write(row, 8, new_regB_rc)\n",
    "                        sheet1.write(row, 9, proteinF_rc)\n",
    "                        sheet1.write(row, 10,proteinB_rc)\n",
    "\n",
    "                self.complete = True \n",
    "                book.save(self.name)\n",
    "                print \"file saved\"\n",
    "                return self.complete\n",
    "                \n",
    "                \n",
    "\n",
    "                \n",
    "                \n",
    "                \n",
    "def main():\n",
    "    plasmidIDer = plasmidIDtool()\n",
    "    plasmidIDer.start()\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main()\n",
    " \n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# /Users/wyssuser/Desktop/testFID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "test_ID = plasmidIDtool()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "enter the folder path you would like to analyze: /Users/wyssuser/Desktop/Repos/Plasmid-Sequence-Identifier/test_seqence\n",
      "Written content in the file successfully\n",
      "Checking files\n",
      "You have 10 file(s)\n",
      "['1-65 (1).seq', '1-65 test again.seq', 'Aspergillus bombycis putative carboxyl ester lipase.fa', 'Drosophila virilis strain 15010-1051.fa', 'Gallus gallus Chicken erythroid transport proteins c1 and c2 .gb', 'Mus musculus cDNA clone MGC.fa', 'Rattus norvegicus G protein-coupled receptor 182.gb', 'sequence1.dna', 'sequence1.fa', 'Xenopus laevis T-box 6 L homeolog.gb']\n",
      "enter the refernce region(s) you would like to search, seperate regions with space: CATAGGCT CATAGGCTCATAGGCT ATCGATCG AAAAA\n",
      "how many base pairs would you like to search : 10\n",
      "how many base pairs is region you like to find : 3\n",
      "what you like to call your excel file : test2\n",
      "1-65 (1).seq\n",
      "Region Found\n",
      "  \n",
      "1-65 test again.seq\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Aspergillus bombycis putative carboxyl ester lipase.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Drosophila virilis strain 15010-1051.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Gallus gallus Chicken erythroid transport proteins c1 and c2 .gb\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Mus musculus cDNA clone MGC.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Rattus norvegicus G protein-coupled receptor 182.gb\n",
      "Region Found\n",
      "  \n",
      "sequence1.dna\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "sequence1.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Xenopus laevis T-box 6 L homeolog.gb\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "1-65 (1).seq\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "1-65 test again.seq\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Aspergillus bombycis putative carboxyl ester lipase.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Drosophila virilis strain 15010-1051.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Gallus gallus Chicken erythroid transport proteins c1 and c2 .gb\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Mus musculus cDNA clone MGC.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Rattus norvegicus G protein-coupled receptor 182.gb\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "sequence1.dna\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "sequence1.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Xenopus laevis T-box 6 L homeolog.gb\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "1-65 (1).seq\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "1-65 test again.seq\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Aspergillus bombycis putative carboxyl ester lipase.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Drosophila virilis strain 15010-1051.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Gallus gallus Chicken erythroid transport proteins c1 and c2 .gb\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Mus musculus cDNA clone MGC.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Rattus norvegicus G protein-coupled receptor 182.gb\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "sequence1.dna\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "sequence1.fa\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "Xenopus laevis T-box 6 L homeolog.gb\n",
      "Region does not exsist in given sequence \n",
      "  \n",
      "1-65 (1).seq\n",
      "Region Found\n",
      "  \n",
      "1-65 test again.seq\n",
      "Region Found\n",
      "  \n",
      "Aspergillus bombycis putative carboxyl ester lipase.fa\n",
      "Region Found\n",
      "  \n",
      "Drosophila virilis strain 15010-1051.fa\n",
      "Region Found\n",
      "  \n",
      "Gallus gallus Chicken erythroid transport proteins c1 and c2 .gb\n",
      "Region Found\n",
      "  \n",
      "Mus musculus cDNA clone MGC.fa\n",
      "Region Found\n",
      "  \n",
      "Rattus norvegicus G protein-coupled receptor 182.gb\n",
      "Region Found\n",
      "  \n",
      "sequence1.dna\n",
      "Region Found\n",
      "  \n",
      "sequence1.fa\n",
      "Region Found\n",
      "  \n",
      "Xenopus laevis T-box 6 L homeolog.gb\n",
      "Region Found\n",
      "  \n",
      "file saved\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_ID.start()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "CATAGGCT CATAGGCTCATAGGCT ATCGATCG AAAAA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    " def ProduceCleanSeq(filename): \n",
    "        #MAKE SURE USELESS IS A CONSTANT \n",
    "        \"\"\"\n",
    "        Format file, removing unessential characters. Useless constant is the character length of \n",
    "        the file for .seq,.fa,.gb,. files.\n",
    "    \n",
    "        \"\"\"\n",
    "       \n",
    "        if filename.endswith('.fa'):\n",
    "            for seq_record in SeqIO.parse(filename, \"fasta\"): Data_formated = str(seq_record.seq).upper()\n",
    "        elif filename.endswith('.gb'):\n",
    "            for seq_record in SeqIO.parse(filename, \"genbank\"): Data_formated = str(seq_record.seq)    \n",
    "        elif filename.endswith('.dna'): \n",
    "            Data_formated = str(snapgene_file_to_dict(filename)['seq']).upper()\n",
    "        elif filename.endswith('.seq'):\n",
    "                useless = 12 \n",
    "                data = open(filename).read().split()\n",
    "                Data_unformatted = list(chain.from_iterable(data[i] for i in range(len(data))))\n",
    "                Data_formated = Data_unformatted[useless:]\n",
    "                \n",
    "        Data_str = ''.join(Data_formated)\n",
    "        Clean_Seq = ''.join([i for i in Data_str if not i.isdigit()])\n",
    "        \n",
    "        return Clean_Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "data = ProduceCleanSeq('sequence1.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GATCCTCCATATACAACGGTATCTCCACCTCAGGTTTAGATCTCAACAACGGAACCATTGCCGACATGAGACAGTTAGGTATCGTCGAGAGTTACAAGCTAAAACGAGCAGTAGTCAGCTCTGCATCTGAAGCCGCTGAAGTTCTACTAAGGGTGGATAACATCATCCGTGCAAGACCAAGAACCGCCAATAGACAACATATGTAACATATTTAGGATATACCTCGAAAATAATAAACCGCCACACTGTCATTATTATAATTAGAAACAGAACGCAAAAATTATCCACTATATAATTCAAAGACGCGAAAAAAAAAGAACAACGCGTCATAGAACTTTTGGCAATTCGCGTCACAAATAAATTTTGGCAACTTATGTTTCCTCTTCGAGCAGTACTCGAGCCCTGTCTCAAGAATGTAATAATACCCATCGTAGGTATGGTTAAAGATAGCATCTCCACAACCTCAAAGCTCCTTGCCGAGAGTCGCCCTCCTTTGTCGAGTAATTTTCACTTTTCATATGAGAACTTATTTTCTTATTCTTTACTCTCACATCCTGTAGTGATTGACACTGCAACAGCCACCATCACTAGAAGAACAGA'"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    " def find_region(Region,clean_seq):\n",
    "        \"\"\"\n",
    "        return bool if region is in sequence\n",
    "        \"\"\"\n",
    "       \n",
    "        return  Region in clean_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def find_region_range(Region,clean_seq,searchRange,searchRegionLen): \n",
    "        #DEAL WITH THE WRAPAROUND ISSUE \n",
    "        \n",
    "        \"\"\"\n",
    "        Look for given region in the cleaned sequence\n",
    "        Foreward is 3' to 5'\n",
    "        \"\"\"\n",
    "        \n",
    "        print len(clean_seq)\n",
    "        \n",
    "        if find_region(Region,clean_seq) == False: \n",
    "            print \"Region does not exsist in given sequence \"\n",
    "            print \"  \"\n",
    "            new_regF, new_regB = \"NaN\",\"NaN\"\n",
    "        \n",
    "        else:\n",
    "            print \"Region Found\"\n",
    "            print \"  \"\n",
    "            regionlen = len(Region)\n",
    "            region_start = clean_seq.find(Region)\n",
    "            region_end = region_start + regionlen\n",
    "            new_pos_F = region_end + searchRange\n",
    "            new_pos_B = region_start - searchRange\n",
    "            \n",
    "            #deal with wrap around issue. \n",
    "            if new_pos_B < 0: \n",
    "                #THINK OF BETTER WAY TO DO THIS\n",
    "                new_pos_B = len(clean_seq) * 2\n",
    "            \n",
    "            new_regF = clean_seq[new_pos_F : new_pos_F + searchRegionLen]\n",
    "            new_regB = clean_seq[new_pos_B - searchRegionLen: new_pos_B]\n",
    "\n",
    "        return new_regF, new_regB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "600\n",
      "Region Found\n",
      "  \n"
     ]
    }
   ],
   "source": [
    "x,y = find_region_range('GATCCTCCATATA',data,1,3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print x\n",
    "print y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
